src/compas/datastructures/tree/tree.py

```diff
@@ -0,0 +1,264 @@
+    children: set[:class:`~compas.datastructures.TreeNode`]
+        The children of the tree node.
+    tree: :class:`~compas.datastructures.Tree`
+        The tree the node belongs to.
+    is_root: bool
+        True if the node is the root node of the tree.
+    is_leaf: bool
+        True if the node is a leaf node of the tree.
+    is_branch: bool
+        True if the node is a branch node of the tree.
+    acestors: generator[:class:`~compas.datastructures.TreeNode`]
+        A generator of the acestors of the tree node.
+    descendants: generator[:class:`~compas.datastructures.TreeNode`]
+        A generator of the descendants of the tree node.
+
+    """
+
+    def __init__(self, name=None, attributes=None):
+        self.name = name
+        self.attributes = attributes or {}
+        self._parent = None
+        self._children = set()
+        self._tree = None
+
+    def __repr__(self):
+        return "<TreeNode {}>".format(self.name)
+
+    @property
+    def is_root(self):
+        return self._parent is None
+
+    @property
+    def is_leaf(self):
+        return not self._children
+
+    @property
+    def is_branch(self):
+        return not self.is_root and not self.is_leaf
+
+    @property
+    def parent(self):
+        return self._parent
+
+    @property
+    def children(self):
+        return self._children
+
+    @property
+    def tree(self):
+        return self._tree
+
+    def add(self, node):
+        """Add a child node to this node."""
+        if not isinstance(node, TreeNode):
+            raise TypeError("The node is not a TreeNode object.")
+        self._children.add(node)
+        node._parent = self
+        node._tree = self.tree
+        if self.tree:
+            self.tree.nodes.add(node)
+
+    def remove(self, node):
+        """Remove a child node from this node."""
+        self._children.remove(node)
+        node._parent = None
+        node._tree = None
+        if self.tree:
+            self.tree.nodes.remove(node)
+
+    @property
+    def ancestors(self):
+        this = self
+        while this:
+            yield this
+            this = this.parent
+
+    @property
+    def descendants(self):
+        for child in self.children:
+            yield child
+            for descendant in child.descendants:
+                yield descendant
+
+    def traverse(self, strategy="preorder"):
+        """Traverse the tree from this node."""
+        if strategy == "preorder":
+            yield self
+            for child in self.children:
+                for node in child.traverse(strategy):
+                    yield node
+        elif strategy == "postorder":
+            for child in self.children:
+                for node in child.traverse(strategy):
+                    yield node
+            yield self
+        else:
+            raise ValueError("Unknown traversal strategy: {}".format(strategy))
+
+
+class Tree(Datastructure):
+    """A tree data structure.
```

Comment:
perhaps a short overview of the most important characteristics of the data structure would be useful.